### ISSUE TITLE
[Bug]: LuaSnip prevent jumping out of snippet when jumping-next-prev

### ISSUE THREAD

**Issue by maxxnino** [LunarVim/LunarVim#1645]
### Problem description

When we don't complete a snippet, exit `insert` mode, open new line with `o`, press `tab`, it's jump to the next location in the old snippet. We can prevent this with `region_check_events` option. But I don't know how to add this. hope someone can fix it

https://github.com/L3MON4D3/LuaSnip/issues/116

https://github.com/L3MON4D3/LuaSnip/issues/170

### LunarVim version

lastest

### Neovim version (>= 0.5)

NVIM v0.6.0-dev+176-g4b3ffde20

### I have

- [X] Read the readme
- [X] Read the wiki
- [X] Searched for similar issues

**Comment by abzcoding** [comment]
can you test if adding this to your config.lua fixes that issue or not?
```lua
local luasnip = require "luasnip"

luasnip.config.set_config {
  history = true,
  -- Update more often, :h events for more info.
  updateevents = "TextChanged,TextChangedI",
  delete_check_events = "TextChanged",
}
```

**Comment by abzcoding** [comment]
if it did work, we can add it as the default config
cc @kylo252 @ChristianChiarulli

**Comment by maxxnino** [comment]
Yes, it did work. But `region_check_events = "CursorMoved"` remove a really nice feature. If we accidently exit `insert` mode in middle of snippet, we want come back to `insert` mode, and press `tab` to continue with the rest of the snippet. It's would be nice if we have some key binding for manual clear all the snippet history

**Comment by abzcoding** [comment]
> Yes, it did work. But `region_check_events = "CursorMoved"` remove a really nice feature. If we accidently exit `insert` mode in middle of snippet, we want come back to `insert` mode, and press `tab` to continue with the rest of the snippet. It's would be nice if we have some key binding for manual clear all the snippet history

maybe try mappaing `<Plug>luasnip-delete-check` to sth and see if that helps
or mapping this `<cmd>lua require'luasnip'.unlink_current_if_deleted()<Cr>`

**Comment by maxxnino** [comment]
```lua
lvim.builtin.which_key.mappings.k = { "<Plug>luasnip-delete-check", "Clear LuaSnip" }
lvim.builtin.which_key.mappings.k = { "<cmd>lua require'luasnip'.unlink_current_if_deleted()<Cr>", "Clear LuaSnip" }
```
I tried both in the config.lua. But it didn't help. I also deleted the setting below

> ```lua
> local luasnip = require "luasnip"
> 
> luasnip.config.set_config {
>   history = true,
>   updateevents = "TextChanged,TextChangedI",
>   delete_check_events = "TextChanged", -- or maybe "InsertLeave"
> --  region_check_events = "CursorMoved", -- or maybe "InsertEnter"
> }
> ```

**Comment by abzcoding** [comment]
> ```lua
> ```lua
> lvim.builtin.which_key.mappings.k = { "<Plug>luasnip-delete-check", "Clear LuaSnip" }
> lvim.builtin.which_key.mappings.k = { "<cmd>lua require'luasnip'.unlink_current_if_deleted()<Cr>", "Clear LuaSnip" }
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> edited: it worked. but do you have setting for the press o for open new line issue
> ```

maybe we need some combination of that setting and using this key mappings 😢

**Comment by abzcoding** [comment]
btw both of those mappings do the same, just wanted to clarify

**Comment by maxxnino** [comment]
yes, both do the same thing. solve only for the delete issue

**Comment by kylo252** [comment]
> if it did work, we can add it as the default config cc @kylo252 @ChristianChiarulli

I sadly have next-to-zero experience with LuaSnip outside of the default behavior that it provides. Make sure it's not conflicting with the changes that @chase added in https://github.com/LunarVim/LunarVim/pull/1523/files#

**Comment by chase** [comment]
It shouldn't conflict, in my config I have:
```
augroup CustomLuaSnip
	au!
	au TextChanged,InsertLeave * lua require'luasnip'.unlink_current_if_deleted()
augroup end
```

I also have:
```
lvim.keys = vim.tbl_deep_extend("force", lvim.keys, {
  insert_mode = {
    ["<C-l>"] = "<C-o>$<cmd>silent! LuaSnipUnlinkCurrent<CR>",
    ["<C-j>"] = "<C-o>o<cmd>silent! LuaSnipUnlinkCurrent<CR>",
...
```

**Comment by abzcoding** [comment]
> It shouldn't conflict, in my config I have:
> 
> ```
> augroup CustomLuaSnip
> 	au!
> 	au TextChanged,InsertLeave * lua require'luasnip'.unlink_current_if_deleted()
> augroup end
> ```
> 
> I also have:
> 
> ```
> lvim.keys = vim.tbl_deep_extend("force", lvim.keys, {
>   insert_mode = {
>     ["<C-l>"] = "<C-o>$<cmd>silent! LuaSnipUnlinkCurrent<CR>",
>     ["<C-j>"] = "<C-o>o<cmd>silent! LuaSnipUnlinkCurrent<CR>",
> ...
> ```

@maxxnino can you check if this solution works for you or not?

**Comment by maxxnino** [comment]
yes, it works really nice, the <c-o> really nice

**Comment by maxxnino** [comment]
yes, it works really nice

**Comment by abzcoding** [comment]
@kylo252 @chase  do we want to add this as the default? or do we better off letting users add this if they like?

**Comment by chase** [comment]
> @kylo252 @chase do we want to add this as the default to lunarvim? or are we better off letting users add this if they like?

I think that the default behavior of LuaSnip is a little unexpected, but the intended behavior may be what people want: to be able to re-enter a snippet even after exiting insert mode.

Ultimately LuaSnip fails to check if the region is still within the bounds after the cursor changes, which I think is a bug. I haven't looked if there is a better way of handling this that is not the default, but in the linked issue, while there is configuration available for LuaSnip I don't see why an explicit bounds check shouldn't be in place.

Maybe along with a default configuration to exit the snippet on InsertLeave, an explicit keymap to jump back into the last snippet would be a more user friendly approach?

**Comment by maxxnino** [comment]
> Maybe along with a default configuration to exit the snippet on InsertLeave, an explicit keymap to jump back into the last snippet would be a more user friendly approach?

yes, that really good idea.

**Comment by abzcoding** [comment]
> > @kylo252 @chase do we want to add this as the default to lunarvim? or are we better off letting users add this if they like?
> 
> I think that the default behavior of LuaSnip is a little unexpected, but the intended behavior may be what people want: to be able to re-enter a snippet even after exiting insert mode.
> 
> Ultimately LuaSnip fails to check if the region is still within the bounds after the cursor changes, which I think is a bug. I haven't looked if there is a better way of handling this that is not the default, but in the linked issue, while there is configuration available for LuaSnip I don't see why an explicit bounds check shouldn't be in place.
> 
> Maybe along with a default configuration to exit the snippet on InsertLeave, an explicit keymap to jump back into the last snippet would be a more user friendly approach?

do you have time to make a PR for it? if not, i'll look into it next week

**Comment by chase** [comment]
> > > @kylo252 @chase do we want to add this as the default to lunarvim? or are we better off letting users add this if they like?
> > 
> > 
> > I think that the default behavior of LuaSnip is a little unexpected, but the intended behavior may be what people want: to be able to re-enter a snippet even after exiting insert mode.
> > Ultimately LuaSnip fails to check if the region is still within the bounds after the cursor changes, which I think is a bug. I haven't looked if there is a better way of handling this that is not the default, but in the linked issue, while there is configuration available for LuaSnip I don't see why an explicit bounds check shouldn't be in place.
> > Maybe along with a default configuration to exit the snippet on InsertLeave, an explicit keymap to jump back into the last snippet would be a more user friendly approach?
> 
> do you have time to make a PR for it? if not, i'll look into it next week

I'm afraid I don't, no.

**Comment by abzcoding** [comment]
> > do you have time to make a PR for it? if not, i'll look into it next week
> 
> I'm afraid I don't, no.

no worries mate, thanks anyway
